gateway/downlink.go

```diff
@@ -0,0 +1,212 @@
+package gateway
+
+/*
+#cgo CFLAGS: -I./sx1302/libloragw/inc -I./sx1302/libtools/inc
+#cgo LDFLAGS: -L./sx1302/libloragw -lloragw -L./sx1302/libtools -lbase64 -lparson -ltinymt32  -lm
+
+#include "../sx1302/libloragw/inc/loragw_hal.h"
+#include "gateway.h"
+#include <stdlib.h>
+
+*/
+import "C"
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/viam-modules/gateway/node"
+	"go.thethings.network/lorawan-stack/v3/pkg/crypto"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+	"go.viam.com/utils"
+)
+
+const (
+	joinDelay     = 6         // rx2 delay in seconds for sending join accept message.
+	downlinkDelay = 2         // rx2 delay in seconds for downlink messages.
+	rx2Frequency  = 923300000 // Frequency to send downlinks on rx2 window
+	rx2SF         = 12        // spreading factor for rx2 window
+	rx2Bandwidth  = 0x06      // 500k bandwidth
+)
+
+func (g *gateway) sendDownlink(ctx context.Context, payload []byte, join bool, packetTime time.Time) error {
+	txPkt := C.struct_lgw_pkt_tx_s{
+		freq_hz:     C.uint32_t(rx2Frequency),
+		freq_offset: C.int8_t(0),
+		// tx_mode 0 is immediate, 1 for timestampted with count_us delay
+		// doing immediate mode with sleep to exit on context cancelation.
+		tx_mode:    C.uint8_t(0),
+		rf_chain:   C.uint8_t(0),
+		rf_power:   C.int8_t(26),            // tx power in dbm
+		modulation: C.uint8_t(0x10),         // LORA modulation
+		bandwidth:  C.uint8_t(rx2Bandwidth), // 500k
+		datarate:   C.uint32_t(rx2SF),
+		coderate:   C.uint8_t(0x01), // code rate 4/5
+		invert_pol: C.bool(true),    // Downlinks are always reverse polarity.
+		size:       C.uint16_t(len(payload)),
+		preamble:   C.uint16_t(8),
+		no_crc:     C.bool(true), // CRCs in uplinks only
+		no_header:  C.bool(false),
+	}
+
+	var cPayload [256]C.uchar
+	for i, b := range payload {
+		cPayload[i] = C.uchar(b)
+	}
+	txPkt.payload = cPayload
+
+	var waitDuration time.Duration
+	switch join {
+	case true:
+		// 47709/32*time.Microsecond is the internal delay of sending a packet
+		waitDuration = (joinDelay * time.Second) - (time.Since(packetTime)) - 47709/32*time.Microsecond
+	default:
+		waitDuration = (downlinkDelay * time.Second) - (time.Since(packetTime)) - 47709/32*time.Microsecond
+	}
+
+	if !accurateSleep(ctx, waitDuration) {
+		return fmt.Errorf("error sending downlink: %w", ctx.Err())
+	}
+
+	errCode := int(C.send(&txPkt))
+	if errCode != 0 {
+		return errors.New("failed to send downlink packet")
+	}
+	// wait for packet to finish sending.
+	var status C.uint8_t
+	for {
+		C.lgw_status(txPkt.rf_chain, 1, &status)
+		if err := ctx.Err(); err != nil {
+			return fmt.Errorf("error sending downlink: %w", ctx.Err())
+		}
+		// status of 2 indicates send was successful
+		if int(status) == 2 {
+			break
+		}
+		time.Sleep(2 * time.Millisecond)
+	}
+
+	g.logger.Infof("sent the downlink packet")
```

Comment:
we should move this log into `uplink.go` and report which device we sent a downlink to. 

worth keeping at info level since these are not spammy